### PR TITLE
apiserver: add --shutdown-delay-duration to keep serving until LBs stop sending traffic

### PR DESCRIPTION
**Pull request by sttts**
This is meant to delay the apiserver shutdown for a defined time duration in order to give the SDN a chance to update changed endpoints.

The reconciler is part of the "master controller", also called "bootstrap controller". It has a pre shutdown hook triggered by the `stopCh`. We delay the `internalStopCh` being closed which triggers to stop serving.

```release-note
Add --shutdown-delay-duration to kube-apiserver in order to delay a graceful shutdown. `/healthz` will keep returning success during this time and requests are normally served, but `/readyz` will return faillure immediately. This delay can be used to allow the SDN to update iptables on all nodes and stop sending traffic.
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go [hunk 1]
```diff
@@ -174,5 +181,9 @@ func (s *ServerRunOptions) AddUniversalFlags(fs *pflag.FlagSet) {
 		"handler, which picks a randomized value above this number as the connection timeout, "+
 		"to spread out load.")
 
+	fs.DurationVar(&s.MinimalShutdownDuration, "minimal-shutdown-duration", s.MinimalShutdownDuration, ""+
+		"Minimal duration of a graceful shutdown, e.g. to guarantee that all endpoints pointing to this API server "+
```
Comment (stewart-yu) [inline staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go]: can can remove `""` at the end of line 184?

File: staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go [hunk 2]
```diff
@@ -119,6 +122,10 @@ func (s *ServerRunOptions) Validate() []error {
 		errors = append(errors, fmt.Errorf("--min-request-timeout can not be negative value"))
 	}
 
+	if s.MinimalShutdownDuration < 0 {
+		errors = append(errors, fmt.Errorf("--minimum-shutdown-duration can not be negative value"))
+	}
+
```
Comment (stewart-yu) [inline staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go]: seem we have some test case about `Validate()` in test file, add the a case in UT better?

File: staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go (hunk 1 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go]: I am just following the style in this file.

File: staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go]: done

**Comment by deads2k** [comment]
This lgtm.  It helps limit an unnecessary race.  It's not foolproof because we cannot be aware of all of our consumers, but it makes it possible to avoid unnecessary dead endpoints.

@kubernetes/sig-api-machinery-misc

**Comment by lavalamp** [comment]
/hold

not that I necessarily disagree, but I think this might be a big enough addition to the surface area that I want to think about it for a second.

**Comment by lavalamp** [comment]
I am confused about the PR description. It seems to assume that there is some external system that tries to kill and start an apiserver at the same-ish time?

Isn't it better to make that system start the new apiserver, then wait, then stop the old apiserver?

If the problem is routes propagating, I'm not convinced this is a good solution as there's no guarantee about how long that takes (is there?).

I do think we should have some sort of "lame duck mode" where we're shutting down but being nice to users. That probably includes:
* Stop being "ready" (but remain "live")
* Reject new long-running requests (watches, proxy, etc)?
* Reject new requests that won't finish by the shutdown time? (esp. writes?)

Goal would be to have as few open connections as possible by the shutdown time, I guess.

I'm not against this change if it's intended to go in that direction. If it's just to paper over a defect in a roll-out system I'm not convinced, I guess.

If we do something like this, I think it would be better to enter this mode on SIGINT and wait for SIGKILL to exit. It seems like it should be possible to get the grace period from the downward API, unless that leads to some weird bootstrapping problem. But even in that case I think it should be documented as to why we chose a command line flag.

Basically, I don't want a new flag if possible :)

**Comment by sttts** [comment]
> * Stop being "ready" (but remain "live")

This is what this change implements.

> * Reject new long-running requests (watches, proxy, etc)?
> * Reject new requests that won't finish by the shutdown time? (esp. writes?)

This could indeed be added.
 
> Goal would be to have as few open connections as possible by the shutdown time, I guess.
> 
> I'm not against this change if it's intended to go in that direction. If it's just to paper over a defect in a roll-out system I'm not convinced, I guess.
> 
> If we do something like this, I think it would be better to enter this mode on SIGINT and wait for SIGKILL to exit. It seems like it should be possible to get the grace period from the downward API, unless that leads to some weird bootstrapping problem. But even in that case I think it should be documented as to why we chose a command line flag.

We have this SIGINT+KILL logic in place ("graceful termination"). But it cuts off the listening port at the moment the SIGINT arrives. This means that the server is still getting requests because load balancers (external and internal endpoints) still route traffic to us. This is a race, this delay can mitigate.

On the other hand, we could get the graceful termination time. But we don't want to wait those 60 seconds either by default, which would slow down any config change rollout by 60 seconds per apiserver.

> Basically, I don't want a new flag if possible :)

We can hard code it to 5 seconds and we are happy. But that's not the general solution either, is it? :-)

**Comment by lavalamp** [comment]
>> Stop being "ready" (but remain "live")

> This is what this change implements.

How does this change affect liveness and readiness (as perceived by kubelet) differently? I wasn't speaking metaphorically, I was talking about literal liveness and readiness as defined by Kubernetes.

**Comment by sttts** [comment]
> How does this change affect liveness and readiness (as perceived by kubelet) differently? I wasn't speaking metaphorically, I was talking about literal liveness and readiness as defined by Kubernetes.

We don't have `/readyz`, but should, in addition to this change. With distinction between readiness and health, we would get:

* kube-apiserver is terminated (SIGTERM, gracefully)
* kube-apiserver unregisters its endpoint from the kubernetes service (external client will stop sending traffic as soon as kube-proxy on each node converges)
* kube-apiserver immediately switch `/readyz` to return an error.
* kube-apiserver keeps serving `/healthz` for the `--minimum-shutdown-duration` of this PR.
* kube-apiserver stops listening on the port, `/healthz` requests will fail as well.
* kube-apiserver keeps serving the active non-long-running requests.
* kube-apiserver waits until no non-long-running request is active anymore, or 60 seconds have passed.
* kube-apiserver terminates (which also cuts off the long-running requests like watches).

In parallel:

* external load balancer use `/readyz` and stop sending traffic when it returns errors.
* internal kube-proxies are updated by the endpoint removal, but need the usually informer lag of time to converge. They might fail to update iptables (because the global iptables lock is taken by another process), and try again 30 seconds later.

Hence, to really guarantee that connection are not refused, we need the minimum-shutdown-duration to overcome

* kube-proxies to update iptables to make the endpoints effective
* external load balancers to ping `/readyz`.

**Comment by lavalamp** [comment]
> We don't have `/readyz`, but should, in addition to this change. With distinction between readiness and health, we would get:

Can we start out by adding a `/readyz`? That will make changes like this much more compelling for me.

I've talked with @logicalhan some about this in the past.

**Comment by sttts** [comment]
/assign @logicalhan 

Anything left for this?

**Comment by sttts** [comment]
@lavalamp talked with @logicalhan extensively about this and readyz. He is working on readyz now, but we need this one in addition. Ok to unhold?

**Review by logicalhan — APPROVED** [review]
I'm okay with adding this as an incremental change.

/lgtm

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 3]
```diff
@@ -155,6 +155,9 @@ type Config struct {
 	// If specified, long running requests such as watch will be allocated a random timeout between this value, and
 	// twice this value.  Note that it is up to the request handlers to ignore or honor this timeout. In seconds.
 	MinRequestTimeout int
+	// MinimalShutdownDuration allows to block shutdown for some time, e.g. until endpoints pointing to this API server
+	// have converged on all node. During this time, the API server keeps serving.
+	MinimalShutdownDuration time.Duration
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/server/config.go]: Can we come up with a clearer name? I'm not sure what "minimal" is adding.

LameDuckDuration? PostShutdownServingDuration? ServeDuringGracePeriod?

Do we need two flags? One to set the period during which it accepts new (short running) requests, and one to set the period during which it rejects new requests but continues to serve old requests? e.g. PostShutdownAcceptRequestsDuration / PostShutdownServeExistingRequestsDuration.

It is a little weird to me that that this is a duration, rather than just waiting for an external system to complete the shutdown process after it confirms that we are no longer in the relevant load balancers.

Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/server/config.go]: Flags are kind of forever so I'd prefer to get it right if possible :/

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/config.go]: not sure we need the second duration. The only use-case I can come up with is that `/healthz` keeps returning 200. But as we are shutting down anyway, the kubelet does not care anymore about `/healthz`.

About an external system: this depends a lot on the environment, whether the load balancers are purely health check based and there is no further communication between the cluster and the LB. That setup has the advantage that it is much simpler, available with every LB, and there is no risk for cyclic dependencies, especially helpful for control plane components.

We might want to support both: pure health check based + some external control.

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/config.go]: Renamed to `--shutdown-delay-period`.

**Review by logicalhan — COMMENTED** [review]
/lgtm

**Review by logicalhan — APPROVED** [review]
/lgtm

**Comment by sttts** [comment]
Rebased.

@lavalamp @logicalhan please cancel the hold here.

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 4]
```diff
@@ -180,6 +180,11 @@ type GenericAPIServer struct {
 	// HandlerChainWaitGroup allows you to wait for all chain handlers finish after the server shutdown.
 	HandlerChainWaitGroup *utilwaitgroup.SafeWaitGroup
 
+	// ShutdownDelayDuration allows to block shutdown for some time, e.g. until endpoints pointing to this API server
+	// have converged on all node. During this time, the API server keeps serving, /healthz will return 200,
+	// a future /readyz will return future.
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go]: ?

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go]: fixed
